interfaces/connection.go

```diff
@@ -0,0 +1,250 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package interfaces
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/snap"
+)
+
+// Connection represents a connection between ConnectedPlug and ConnectedSlot
+type Connection struct {
+	plug *ConnectedPlug
+	slot *ConnectedSlot
+}
+
+// ConnectedPlug represents a plug that is connected to a slot.
+type ConnectedPlug struct {
+	plugInfo     *snap.PlugInfo
+	staticAttrs  map[string]interface{}
+	dynamicAttrs map[string]interface{}
+}
+
+// ConnectedSlot represents a slot that is connected to a plug.
+type ConnectedSlot struct {
+	slotInfo     *snap.SlotInfo
+	staticAttrs  map[string]interface{}
+	dynamicAttrs map[string]interface{}
+}
+
+// NewConnectedSlot creates an object representing a connected slot.
+func NewConnectedSlot(slot *snap.SlotInfo, dynamicAttrs map[string]interface{}) *ConnectedSlot {
+	return &ConnectedSlot{
+		slotInfo:     slot,
+		staticAttrs:  copyAttributes(slot.Attrs),
+		dynamicAttrs: dynamicAttrs,
+	}
+}
+
+// NewConnectedPlug creates an object representing a connected plug.
+func NewConnectedPlug(plug *snap.PlugInfo, dynamicAttrs map[string]interface{}) *ConnectedPlug {
+	return &ConnectedPlug{
+		plugInfo:     plug,
+		staticAttrs:  copyAttributes(plug.Attrs),
+		dynamicAttrs: dynamicAttrs,
+	}
+}
+
+// Interface returns the name of the interface for this plug.
+func (plug *ConnectedPlug) Interface() string {
+	return plug.plugInfo.Interface
+}
+
+// Name returns the name of this plug.
+func (plug *ConnectedPlug) Name() string {
+	return plug.plugInfo.Name
+}
+
+// Snap returns the snap Info of this plug.
+func (plug *ConnectedPlug) Snap() *snap.Info {
+	return plug.plugInfo.Snap
+}
+
+// Apps returns all the apps associated with this plug.
+func (plug *ConnectedPlug) Apps() map[string]*snap.AppInfo {
+	return plug.plugInfo.Apps
```

Comment:
Perhaps ok but I worry about apps being modified because the caller got them through this method.

Reply:
the problem cannot be worse than now where we have PlugInfo.Apps